### ISSUE TITLE
Transfers of large files occasionally very slow

### ISSUE THREAD

**Issue by davidgubler** [grke/burp#874]
burp appears to have an issue where large file transfers are sometimes very slow (only ~400 KB/s). This happens sporadically (like every 10th - 100th time) on some systems where burp usually works fine. There is no obvious trigger (configuration change, larger than usual files or similar) when this happens.

We've had this issue for a while now with older burp versions. We're currently experiencing it with 2.3.30 (ubuntu package by VSHN) on Ubuntu bionic (both server and client). We're still using protocol version 1.

Unfortunately it is very difficult to point out exactly what goes on, but here's what I've found when the issue appears:
* Happens very rarely, without configuration change or other substantial changes on the server or client
* When it happens it is very pronounced, client only transfers data at a pretty stable 400 KB/s over a network that usually does 50 MB/s or more
* server is in "Begin phase2 (receive file data)"
* Happens with large single files (e.g. 73 GB Gitlab dump on an otherwise normal system)
* CPU load and IO wait on client and server are near zero
* Network bandwidth is not a limiting factor, no per-connection traffic shaping or similar in place. We've done various performance tests when this issue is in progress, and found that the involved machines perform just fine (disk, network, cpu). While burp transmits data at 400 KB/s it is easily possible to copy files at >50 MB/s over scp at the same time.
* The client is quickly switching between write() and select() system calls: burp_client_strace.txt (last column is time taken). Judging by the timing of the strace output, the select() calls seem to be limiting the transfer rate
* The server has a lot of EAGAIN on fd 7 (presumably that's the socket): burp_server_strace.txt (last column is time taken). This may be normal though. Also occasionally there are select() calls that take several ms.
* TCP queues ("ss -6"): Send queue very large on the client (~150 kb), receive queue not empty on the client (42 bytes). Both queues are empty most of the time on the server.

An interesting aspect is that the TCP receive buffer on the client always contains 42 bytes (it stays on that over hours). This may indicate that the client does not read from the TCP socket in a timely manner.

I'm just guessing wildly here:
It kind of looks like the protocol code on the client and server aren't exactly in sync anymore and the client doesn't do what the server expects (and the other way round). Like: The client sends data when actually it should be reading statistics from the socket, or something like that.

We have ~1000 systems using burp, and we only experience this like once a week, thus it isn't exactly a huge problem, but still repeatedly requires manual intervention. It mostly seems to affect certain systems, in particular Gitlabs with their large single-file dumps.

[burp_client_strace.txt](https://github.com/grke/burp/files/5402767/burp_client_strace.txt)
[burp_server_strace.txt](https://github.com/grke/burp/files/5402769/burp_server_strace.txt)

**Comment by grke** [comment]
Hello,

It looks to me like something has gone wrong on the server side and it is no longer reading properly.

Are you using the config librsync=0? (librsync=1 won't work well with files that huge).
Are you using the burp status monitor (or burp-ui) on the server?
Does the server side log anything in the backup log file when this happens?

**Comment by davidgubler** [comment]
> Are you using the config librsync=0?

We do not configure that option, hence it seems to be enabled. I'll try to disable it on a few systems. However it may take days or weeks to find out if it has an effect.

> burp-ui

I think we have it somewhere, I'll check next time I see the problem.

> Does the server side log anything in the backup log file when this happens?

No. The last line is "Begin phase2 (receive file data)".

**Comment by grke** [comment]
If you have huge files that get touched, librsync=1 is probably your problem.
The first backup of such a file will be OK, but then librsync will take forever to generate signatures and diffs.
The disadvantage of librsync=0 is that the huge file will be backed up a second time if it is touched, but librsync won't lock everything up.
I recommend ruling that out before anything else. Doesn't have to take days or weeks, just backup a huge file, then touch it, then try backing it up again.

There is another config that you can think about if you like librsync being on - check 'librsync_max_size' in the man page.

You may also find the 'exclude_logic'+'file_size' options interesting.

**Comment by davidgubler** [comment]
Hi,

I've now tested both "librsync_max_size = 1GB" and "librsync = 0" with one of our most affected servers by putting these settings into /etc/burp/clients/<hostname> (which I hope is the right thing to do), but it didn't help at all, the problem quickly reappeared both times. On this one server the problem appears ca. every 10th backup.

One thing I noticed though after tracking the occurrences more carefully is that while we have a variety of clients that are affected (not only gitlabs but also database servers, which makes sense I guess), they mostly seem to use one out of six burp servers. That one burp server isn't special though, not in terms of hardware nor in terms of software or configuration, and doesn't show any signs of hardware issues... (there's also been one instance that happened with a different burp server, but I'm not entirely sure if it's the same problem).
